### PR TITLE
Introduce `maintenance.gardener.cloud/operation` annotation for `Shoot`s

### PR DESCRIPTION
**Pull request by rfranzke**
**How to categorize this PR?**

/area usability
/kind enhancement

**What this PR does / why we need it**:
While it is possible to confine `spec` update rollout to the maintenance time window, we do not offer any option to confine operation rollouts.

This PR introduces a new `maintenance.gardener.cloud/operation` annotation for `Shoot`s which can be used to confine the execution of the respective operation to the shoot cluster's maintenance time window.

**Release note**:

```feature user
With the new `maintenance.gardener.cloud/operation` annotation for `Shoot`s it is now possible to confine the execution of the respective operation to the shoot cluster's maintenance time window.
```

### CODE REVIEW

File: pkg/apis/core/v1beta1/helper/helper.go [hunk 1]
```diff
@@ -1660,3 +1660,34 @@ func MutateShootETCDEncryptionKeyRotation(shoot *gardencorev1beta1.Shoot, f func
...
+		if operation == v1beta1constants.ShootOperationRetry {
+			return true, true
+		}
+	} else {
+		switch operation {
+		case v1beta1constants.GardenerOperationReconcile:
+			return true, true
+
+			// TODO(rfranzke): Handle new operation annotations introduced by
+			// - https://github.com/gardener/gardener/pull/6021
```
Comment (ialidzhikov) [inline pkg/apis/core/v1beta1/helper/helper.go]:
```suggestion
```

#6021 is merged, we can add to the list the operation annotations related to the etcd encryption key.

File: pkg/controllermanager/controller/shoot/shoot_maintenance_control.go [hunk 2]
```diff
@@ -378,6 +378,31 @@ func hasMaintainNowAnnotation(shoot *gardencorev1beta1.Shoot) bool {
 	return ok && operation == v1beta1constants.ShootOperationMaintain
 }
 
+func needsRetry(shoot *gardencorev1beta1.Shoot) bool {
+	needsRetryOperation := false
+
+	if val, ok := shoot.Annotations[v1beta1constants.FailedShootNeedsRetryOperation]; ok {
+		needsRetryOperation, _ = strconv.ParseBool(val)
+	}
+	delete(shoot.Annotations, v1beta1constants.FailedShootNeedsRetryOperation)
```
Comment (ialidzhikov) [inline pkg/controllermanager/controller/shoot/shoot_maintenance_control.go]: The func is called `needsRetry` but it also has a side effect and deletes the corresponding annotation. Maybe the can be moved out of this func or the name can be adjusted.

File: pkg/controllermanager/controller/shoot/shoot_maintenance_control.go [hunk 3]
```diff
@@ -378,6 +378,31 @@ func hasMaintainNowAnnotation(shoot *gardencorev1beta1.Shoot) bool {
...
+func getOperation(shoot *gardencorev1beta1.Shoot) string {
+	var (
+		operation            = v1beta1constants.GardenerOperationReconcile
+		maintenanceOperation = shoot.Annotations[v1beta1constants.GardenerMaintenanceOperation]
+	)
+
+	if isValid, _ := gardencorev1beta1helper.IsValidShootOperation(maintenanceOperation, shoot.Status.LastOperation); isValid {
+		operation = maintenanceOperation
+	}
+	delete(shoot.Annotations, v1beta1constants.GardenerMaintenanceOperation)
```
Comment (ialidzhikov) [inline pkg/controllermanager/controller/shoot/shoot_maintenance_control.go]: It should be a side-effect as well. Probably we should rename the func or move the deletion outside.

File: pkg/controllermanager/controller/shoot/shoot_maintenance_control.go [hunk 4]
```diff
@@ -378,6 +378,31 @@ func hasMaintainNowAnnotation(shoot *gardencorev1beta1.Shoot) bool {
...
+	return needsRetryOperation
+}
+
+func getOperation(shoot *gardencorev1beta1.Shoot) string {
+	var (
+		operation            = v1beta1constants.GardenerOperationReconcile
+		maintenanceOperation = shoot.Annotations[v1beta1constants.GardenerMaintenanceOperation]
+	)
+
+	if isValid, _ := gardencorev1beta1helper.IsValidShootOperation(maintenanceOperation, shoot.Status.LastOperation); isValid {
```
Comment (ialidzhikov) [inline pkg/controllermanager/controller/shoot/shoot_maintenance_control.go]: IIUC the second return value `removedBeforePersisted` is not used. Is it a leftover?

File: pkg/apis/core/v1beta1/helper/helper.go [hunk 5]
```diff
@@ -1660,3 +1660,34 @@ func MutateShootETCDEncryptionKeyRotation(shoot *gardencorev1beta1.Shoot, f func
...
+			// once those PRs got merged.
+		case v1beta1constants.ShootOperationRotateKubeconfigCredentials,
+			v1beta1constants.ShootOperationRotateObservabilityCredentials,
+			v1beta1constants.ShootOperationRotateSSHKeypair,
+			v1beta1constants.ShootOperationRotateCAStart,
+			v1beta1constants.ShootOperationRotateCAComplete,
+			v1beta1constants.ShootOperationRotateServiceAccountKeyStart,
+			v1beta1constants.ShootOperationRotateServiceAccountKeyComplete:
+			return true, false
+		}
```
Comment (ialidzhikov) [inline pkg/apis/core/v1beta1/helper/helper.go]: @oliver-goetz had the suggestion to have a validation on apiserver side for the valid/supported operation annotation values. From user's perspective you are now allowed to add operation annotation value that is not valid.

Comment (rfranzke) [reply pkg/apis/core/v1beta1/helper/helper.go]: Good point, I think it makes sense given hat the operation annotations are quite important. However, I think it makes sense to tackle this outside of this PR. @oliver-goetz could you please create an issue for the topic?

File: pkg/controllermanager/controller/shoot/shoot_maintenance_control.go (hunk 4 above)
Comment (rfranzke) [reply pkg/controllermanager/controller/shoot/shoot_maintenance_control.go]: Yeah, it's a left-over. I remove it for now. Initially, I intended to use this function also in `core/shoot/strategy.go`, but didn't do it for now.

**Review by ialidzhikov — COMMENTED** [review]
/lgtm

**Pull request merged**